---
title: "Differential Expression on Mif NaCl Experiment"
author: "Micah Gearhart"
date: "8/19/2016"
output: md_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Libraries
```{r libraries,eval=T,results='hide',message=FALSE,warning=FALSE}
library(Rsamtools)
library(GenomicAlignments)
library(BiocParallel)
library(DESeq2)
library(ggplot2)
library(magrittr)
library(biomaRt)
library(ChIPpeakAnno)
library(Mus.musculus)
library(goseq)
library(ComplexHeatmap)
library(gridExtra)

ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```


## Get annotation data from biomaRt
```{r biomaRt,eval=F}
ensembl_84<-useMart(biomart="ENSEMBL_MART_ENSEMBL",host = "useast.ensembl.org", dataset="mmusculus_gene_ensembl")
txdb84<-makeTxDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",host = "www.ensembl.org",dataset="mmusculus_gene_ensembl")
ens84<-exonsBy(txdb84,by="gene")
mgi <-  getBM(attributes = c("ensembl_gene_id", "mgi_id", "mgi_symbol"), filters = "ensembl_gene_id",
    values = names(ens84), mart = ensembl_84)
save(ens84,file="ens84_mouse.rdata")
save(mgi,file="mgi.rdata")

#Need TSS data for ChIP Annotation
tss<-getAnnotation(ensembl_84,featureType="TSS")
save(tss,file="tss.rdata")

```
## Summarize Overlaps

Run summarizeOverlaps.Rmd on HPC cluster. 

## EDA
```{r}
load("igarash0_Project_004_genehits_ens84Thu_Jun_30_2016_2351.rdata")
load("mgi.rdata")
cds <- DESeqDataSet(genehits,design=~1)
cds$filename<-rownames(colData(cds))
colData(cds)$genotype<-as.factor(sapply(strsplit(rownames(colData(cds)),"-"),function(x){x[1]}))
colData(cds)$salt<-factor(ifelse(sapply(strsplit(rownames(colData(cds)),"-"),function(x){x[2]})=="Nacl","Nacl","noSalt"),levels=c("noSalt","Nacl"))

#PCA plot by Group
plotPCA(DESeqTransform(cds),intgroup=c("salt","genotype"))+
  scale_color_manual(values=cbPalette)  + theme_bw()
```

```{r}
design(cds)<- ~ genotype + salt + genotype:salt
cds <- DESeq(cds)

#Plot Hnf1b Counts
plotCounts(cds,"ENSMUSG00000020679",intgroup=c("salt","genotype"),returnData=F)
```

## ggplot version of plotCounts to facet by genotype
```{r}
gg_plotCounts<-function(x="ENSMUSG00000020679") {
  if (substr(x,1,7)=="ENSMUSG") {
  title<-mgi[grep(x,mgi$ensembl_gene_id),"mgi_symbol"]
  } else {
    title<-x
    x<-mgi[grep(paste0("^",title,"$"),mgi$mgi_symbol),"ensembl_gene_id"]
  }
  
plotCounts(cds,x,intgroup=c("salt","genotype"),returnData=T) %>% 
  ggplot(aes(x=salt, y=count)) +
  geom_point(position=position_jitter(w=0.1,h=0)) + ggtitle(paste0(x,"\n",title)) +
 # scale_y_log10(breaks=c(25,100,400)) + 
 # scale_y_continuous(trans="log2") + ylim(0,25000) +
  expand_limits(x=0, y = 0) +
  facet_grid(~genotype) + theme_bw()
  }

gg_plotCounts("Nr1h4")
```

## Look for genes with High Interaction Term
```{r}
resultsNames(cds)
res<-results(cds, name="genotypeMif.saltNacl",alpha=0.05,lfcThreshold = 1)
summary(res)
resDF<-as.data.frame(res)
resDF<-resDF[!is.na(resDF$padj),]
resDF<-resDF[with(resDF,order(-log2FoldChange)),]
idx<-match(rownames(resDF),mgi$ensembl_gene_id)
resDF$mgi<-mgi[idx,"mgi_symbol"]
resDF<-resDF[with(resDF,order(padj)),]
head(resDF)
resDF["ENSMUSG00000047638",]
g1<-gg_plotCounts("Slc14a1")
g2<-gg_plotCounts("Lrrn3")
tail(resDF)
g3<-gg_plotCounts("Saa3")
g4<-gg_plotCounts("Zfp750")
grid.arrange(g1,g2,g3,g4,ncol=2)

g1<-gg_plotCounts("Ppbp")
g2<-gg_plotCounts("Nr1h4")
g3<-gg_plotCounts("Pkhd1")
g4<-gg_plotCounts("Kdm7a")
grid.arrange(g1,g2,g3,g4,ncol=2)

## Make a list of genes that have high (logFC > 1) Interaction Terms
length(interactive_genes<-rownames(resDF[abs(resDF$log2FoldChange) > 1 & resDF$padj < 0.05,]))

```

# Calculate LogFC within each Genotype
## Control Cells
```{r}
cds2<-cds
cds2$group <- factor(paste0(cds2$genotype, cds2$salt))
design(cds2)<- ~ group
cds2 <- DESeq(cds2) 
resultsNames(cds2)

summary(res_wt<-results(cds2, contrast=c("group","ControlNacl","ControlnoSalt"),alpha=0.05,lfcThreshold = 0))
res_wt<-as.data.frame(res_wt)
colnames(res_wt)<-paste0("WT",".",colnames(res_wt))

#table(ifelse((res_wt$WT.log2FoldChange>1), "up", ifelse((res_wt$WT.log2FoldChange< -1 ), "down", "no_change")))
#table(ifelse((res_wt$WT.log2FoldChange>1 & res_wt$WT.padj < 0.05), "up", 
#             ifelse((res_wt$WT.log2FoldChange< -1 & res_wt$WT.padj < 0.05), "down", "no_change")))
```

## Hnf1b DN Cells
```{r}
summary(res_hnf1b<-results(cds2, contrast=c("group","MifNacl","MifnoSalt"),alpha=0.05,lfcThreshold = 0))
res_hnf1b<-as.data.frame(res_hnf1b)
colnames(res_hnf1b)<-paste0("hnf1b",".",colnames(res_hnf1b))

#table(ifelse(res_hnf1b$hnf1b.log2FoldChange>1, "up", ifelse(res_hnf1b$hnf1b.log2FoldChange< -1, "down", "no_change")))
#table(ifelse((res_hnf1b$hnf1b.log2FoldChange>1 & res_hnf1b$hnf1b.padj < 0.05), "up", 
#             ifelse((res_hnf1b$hnf1b.log2FoldChange< -1 & res_hnf1b$hnf1b.padj < 0.05), "down", "no_change")))

```

## Merge WT and Hnf1b DN Results
```{r}
res_merged<-cbind(res_wt,res_hnf1b)
head(res_merged)

#res_merged<-res_merged[res_merged$WT.padj < 0.05 & res_merged$WT.baseMean > 100,]
res_merged<-subset(res_merged,!(is.na(WT.log2FoldChange) | is.na(hnf1b.log2FoldChange)))
dim(res_merged)
idx<-match(rownames(res_merged),mgi$ensembl_gene_id)
res_merged$mgi<-mgi[idx,"mgi_symbol"]

#Highlight genes with WTlogFC > 1 & strong interaction term
summary(idx<-match(rownames(res_merged),rownames(resDF)))
res_merged$int<-resDF[idx,"log2FoldChange"]
res_merged[is.na(res_merged$int),"int"]<-0
res_merged$int.padj<-resDF[idx,"padj"]

res_merged$iTerm<-abs(res_merged$int) > 1 & res_merged$int.padj < 0.05 & abs(res_merged$WT.log2FoldChange) > 1 & res_merged$WT.padj < 0.05

#plot(res_merged$WT.log2FoldChange,res_merged$hnf1b.log2FoldChange,pch=16,cex=0.5,
#     xlab="WT Salt Induced Log2FC",
#     ylab="Hnf1b Salt Induced Log2FC",
#     main="Log2FC in Hnf1b DN vs WT Cells",
#     col=ifelse(res_merged$iTerm,"red","black"))

#res_merged[identify(res_merged$WT.log2FoldChange,res_merged$hnf1b.log2FoldChange,labels=res_merged$mgi),]
```

#Figure 3C
```{r}
(g1<-ggplot(res_merged,aes(x=WT.log2FoldChange,y=hnf1b.log2FoldChange)) + geom_point(aes(colour = int)) + theme_bw() +
  scale_colour_gradient2("Hnf1b:NaCl\nInteraction\nEffect Size") )

ggsave(paste0("Figure3C_scatterPlot_",ts,".jpg"),g1,width=7,height=7,dpi=600)
```

#Use Published ChIP-Seq data to find out which of these targets might be direct or indirect
## Use macs2 to call peaks

```{bash MACS,eval=F}
macs2 callpeak --call-summits -c igg_rep1_SRR2124926.dedup.unique.bam  -t hnf1b_rep1_SRR2124924.dedup.unique.bam -n hnf1b_rep1 -g mm
macs2 callpeak --call-summits -c igg_rep2_SRR2124927.dedup.unique.bam  -t hnf1b_rep2_SRR2124925.dedup.unique.bam -n hnf1b_rep2 -g mm
```

## Import narrowPeak Files
```{r}
narrowPeakToGRanges<-function(file) {
  x <- read.table(file,stringsAsFactors=F)
  gr <-GRanges(seqnames=x$V1, ranges = IRanges(start=x$V2, end=x$V3),
               strand="*", score=x$V5, e=x$V7,summit=x$V10)
  return(gr)
}

hnf1b_rep1<-narrowPeakToGRanges("hnf1b_rep1_peaks.narrowPeak")
hnf1b_rep2<-narrowPeakToGRanges("hnf1b_rep2_peaks.narrowPeak")
pairs<-findOverlapPairs(hnf1b_rep1,hnf1b_rep2)
hnf1b<-pintersect(pairs)
head(hnf1b)
```

## Annotate Peaks
```{r}
#tss<-getAnnotation(ensembl_84,featureType="TSS")
#save(tss,file="tss.rdata")
load("tss.rdata")

#Subset to genes that are expressed
x<-rownames(res_merged[res_merged$WT.baseMean > 1,])
tss<-tss[names(tss) %in% x]

hnf1b_anno <- annotatePeakInBatch(hnf1b, AnnotationData=tss, output="both", maxgap=100L)
summary(hnf1b_anno$shortestDistance)
table(hnf1b_anno$insideFeature)
p1<-hnf1b_anno[hnf1b_anno$insideFeature=="upstream" & hnf1b_anno$shortestDistance < 50000,]$feature
p2<-hnf1b_anno[hnf1b_anno$insideFeature=="inside" | hnf1b_anno$insideFeature=="overlapStart" | hnf1b_anno$insideFeature=="overlapEnd"]$feature
p<-unique(c(p1,p2))
res_merged$`Nearby Peak`<-rownames(res_merged) %in% p
sum(res_merged$`Nearby Peak`)
```

## Create Output
```{r}
colnames(res_merged)
temp<-res_merged[,c("mgi","WT.baseMean","WT.log2FoldChange","WT.padj","hnf1b.log2FoldChange","hnf1b.padj",
                    "iTerm","int","int.padj","Nearby Peak")]
colnames(temp)<-c("MGI Symbol","baseMean","WT.log2FoldChange","WT.padj","Hnf1b.log2FoldChange","Hnf1b.padj",
    "Interaction","Interaction.Size","Interaction.padj","Nearby Peak")
temp<-temp[with(temp,order(-Interaction.Size)),]
dim(temp)
head(temp)
table(ifelse((temp$WT.log2FoldChange>1 & temp$WT.padj < 0.05), "up", 
             ifelse((temp$WT.log2FoldChange< -1 & temp$WT.padj < 0.05), "down", "no_change")))
table(ifelse((temp$Hnf1b.log2FoldChange>1 & temp$Hnf1b.padj < 0.05), "up", 
             ifelse((temp$Hnf1b.log2FoldChange< -1 & temp$Hnf1b.padj < 0.05), "down", "no_change")))
write.csv(temp,file=paste0("hnf1b_nacl_log2FCs_wChIPdata_",ts,".csv"),quote=F)
```

## Upload Output to googlesheets
```{r,eval=F}
options(httr_oob_default=TRUE)
gs_auth(new_user = TRUE)
temp_gs<-googlesheets::gs_upload(paste0("hnf1b_nacl_log2FCs_wChIPdata_",ts,".csv")
#googlesheets::gs_browse(temp_gs)
```

## GO Analysis of NaCl dependent Genes in WT Cells
```{r}
expressed_genes<-rownames(res_wt[res_wt$WT.baseMean > 1 & !is.na(res_wt$WT.log2FoldChange) & !is.na(res_wt$WT.padj),])
gocat<-AnnotationDbi::select(Mus.musculus,keys=expressed_genes,keytype="ENSEMBL",columns="GOID")
gocat<-gocat[gocat$ONTOLOGY=="BP",c("ENSEMBL","GOID")]
gocat$GOID<-as.character(gocat$GOID)
gocat.list<-split(gocat$GOID,gocat$ENSEMBL)
gocat.list[["ENSMUSG00000047638"]] #fxr
gocat.list[["ENSMUSG00000030109"]] #Slc6a12
gocat.list[["ENSMUSG00000043760"]] #Pkhd1

#bias data
bd<-sum(width(reduce(ranges(genehits))))
bd["ENSMUSG00000047638"]
#bd<-bd[names(bd) %in% expressed_genes]
head(bd<-bd[expressed_genes])

#bias data
temp<-res_wt[expressed_genes,]
degs<-as.numeric(temp$WT.padj < 0.05 & abs(temp$WT.log2FoldChange ) > 1)
names(degs)<-rownames(temp)
table(degs)
summary(degs)

listGO<-function(goid) {
print(OrganismDbi::select(Mus.musculus,keys=goid,keytype="GOID",columns="TERM"))
tg<-OrganismDbi::select(Mus.musculus,keys=gocat[grep(goid,gocat$GOID),"ENSEMBL"],keytype="ENSEMBL",columns="SYMBOL")
tg$deg<-degs[tg$ENSEMBL]
tg
}

pwf<-nullp(degs,bias.data=bd)
GO.wall<-goseq(pwf,gene2cat=gocat.list)

head(GO.wall,20) %>%
  dplyr::mutate(term=factor(term,levels=rev(term))) %>%
ggplot(aes(x=term,y=-log10(over_represented_pvalue))) +
  geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") +
  theme_bw() 


```


## Compare Wt vs Hnf1b DN expression profiles in the absence of Salt
```{r}
summary(res_genotype<-results(cds2, contrast=c("group","MifnoSalt","ControlnoSalt"),alpha=0.05,lfcThreshold = 1))
colnames(res_genotype)<-paste0("Wt_vs_DN",".",colnames(res_genotype))
#head(res_genotype<-as.data.frame(res_genotype))
#head(res_genotype[!(is.na(res_genotype$Wt_vs_DN.padj) | is.na(res_genotype$Wt_vs_DN.log2FoldChange)),])
res_genotype<-res_genotype[!(is.na(res_genotype$Wt_vs_DN.padj) | is.na(res_genotype$Wt_vs_DN.log2FoldChange)),]
idx<-match(rownames(res_genotype),mgi$ensembl_gene_id)
res_genotype$mgi<-mgi[idx,]$mgi_symbol
res_genotype<-res_genotype[with(res_genotype,order(Wt_vs_DN.padj,Wt_vs_DN.log2FoldChange)),]
head(res_genotype)
res_genotype$`Nearby Peak`<-rownames(res_genotype) %in% p

temp<-res_genotype[,c("mgi","Wt_vs_DN.baseMean","Wt_vs_DN.log2FoldChange","Wt_vs_DN.padj","Nearby Peak")]
gg_plotCounts("Pde4c")
write.csv(temp,file=paste0("Wt_vs_DN_log2FCs_wChIPdata_",ts,".csv"),quote=F)
```

## Output to Google Sheets
```{r eval=F}
options(httr_oob_default=TRUE)
googlesheets::gs_auth(new_user = TRUE)
temp_gs<-googlesheets::gs_upload(paste0("Wt_vs_DN_log2FCs_wChIPdata_",ts,".csv"))
#googlesheets::gs_browse(temp_gs)
```

## Figure 3D Heatmap
```{r}
x<-res_merged[res_merged$iTerm & res_merged$`Nearby Peak`,
                 c("mgi","WT.baseMean","WT.log2FoldChange","WT.padj","hnf1b.log2FoldChange","hnf1b.padj","Nearby Peak","int","int.padj")]
colnames(x)<-c("MGI Symbol","baseMean","WT.log2FoldChange","WT.padj","Hnf1b.log2FoldChange","Hnf1b.padj","Nearby Peak","Interaction","Interaction.Padj")

#Get FPKM values for Genes in x
f<-as.data.frame(fpkm(cds2))

remove_X <- function(s) {substr(s,1,nchar(s)-2)}

colnames(f)<-sapply(strsplit(colnames(f),"_"),function(x) x[1]) %>% gsub("-","_",.) 

#Export a Copy of FPKM table for Supplement
temp<-f
colnames(temp)<-gsub("Mif","Hnf1bDN",colnames(temp))
idx<-match(rownames(temp),mgi$ensembl_gene_id)
temp$'MGI Symbol' <- mgi[idx,]$mgi_symbol
temp<-temp[,c(13,1:12)]
write.csv(temp,file=paste0("FPKM_data_",ts,".csv"),quote=F)
rm(temp)

#subset for heatmap
f<-f[rownames(f) %in% rownames(x),]

# dplyr to get mean fpkm per condition
f<-as.data.frame(f) %>% 
  cbind(Gene = rownames(f)) %>% 
  tidyr::gather(Sample,Value,Control_A:Mif_Nacl_C) %>% 
  dplyr::mutate(Sample=remove_X(Sample)) %>% 
  dplyr::group_by(Sample,Gene) %>% dplyr::summarise(fpkm=log2(mean(Value))) %>% 
  tidyr::spread(Sample,fpkm)

#Add labels from MGI
idx<-match(f$Gene,mgi$ensembl_gene_id)
f$label <- mgi[idx,]$mgi_symbol

idx<-match(f$Gene,rownames(res_merged))
f$int<-res_merged[idx,]$int

#f$Effect<-factor(ifelse(f$int < 0,"Failure to Up Regulate","Failure to Down Regulate"),levels=c("Failure to Up Regulate","Failure to Down Regulate"))

f<-f[with(f,order(int)),]

#Alternative sort by WT Log2FC
#f$log2fc<-f$Control_Nacl-f$Control
#f<-f[with(f,order(-log2fc)),]

f2<-as.matrix(f[,2:5])
rownames(f2)<-f$label

ha <- HeatmapAnnotation(Effect_Size = row_anno_barplot(f$int, baseline = 0, gp = gpar(fill = ifelse(f$int > 0, "red", "green")),axis = TRUE),which='row',width = unit(2, "cm"),name="Interaction\nEffect\nSize")

h<-Heatmap(f2,clustering_distance_rows = "pearson",cluster_rows=FALSE,
           cluster_columns = FALSE,name="Mean\nLog2(FPKM)")

h+ha

#split=f$Effect
pdf(paste0("Figure3D_heatmap_",ts,".pdf"),width=6,height=5)
h+ha
dev.off()
```

# SessionInfo 
```{r}
sessionInfo()
```

