---
title: "Differential Expression on Mif NaCl Experiment"
author: "Micah Gearhart"
date: "7/3/2016"
output: md_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Libraries
```{r libraries,eval=T,results='hide',message=FALSE,warning=FALSE}
library(Rsamtools)
library(GenomicAlignments)
library(BiocParallel)
library(DESeq2)
library(ggplot2)
library(magrittr)
library(biomaRt)
library(ChIPpeakAnno)
library(Mus.musculus)
library(goseq)

ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```


## Get annotation data from biomaRt
```{r biomaRt,eval=F}
ensembl_84<-useMart(biomart="ENSEMBL_MART_ENSEMBL",host = "useast.ensembl.org", dataset="mmusculus_gene_ensembl")
txdb84<-makeTxDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",host = "www.ensembl.org",dataset="mmusculus_gene_ensembl")
ens84<-exonsBy(txdb84,by="gene")
mgi <-  getBM(attributes = c("ensembl_gene_id", "mgi_id", "mgi_symbol"), filters = "ensembl_gene_id",
    values = names(ens84), mart = ensembl_84)
save(ens84,file="ens84_mouse.rdata")
save(mgi,file="mgi.rdata")
```

## EDA
```{r}
load("igarash0_Project_004_genehits_ens84Thu_Jun_30_2016_2351.rdata")
load("mgi.rdata")
cds <- DESeqDataSet(genehits,design=~1)
cds$filename<-rownames(colData(cds))
colData(cds)$genotype<-as.factor(sapply(strsplit(rownames(colData(cds)),"-"),function(x){x[1]}))
colData(cds)$salt<-factor(ifelse(sapply(strsplit(rownames(colData(cds)),"-"),function(x){x[2]})=="Nacl","Nacl","noSalt"),levels=c("noSalt","Nacl"))
plotPCA(DESeqTransform(cds),intgroup=c("salt","genotype"))+
  scale_color_manual(values=cbPalette)  + theme_bw()
```

```{r}
design(cds)<- ~ genotype + salt + genotype:salt
cds <- DESeq(cds)

plotCounts(cds,"ENSMUSG00000020679",intgroup=c("salt","genotype"),returnData=F)
```

```{r}
gg_plotCounts<-function(x="ENSMUSG00000020679") {
  if (substr(x,1,7)=="ENSMUSG") {
  title<-mgi[grep(x,mgi$ensembl_gene_id),"mgi_symbol"]
  } else {
    title<-x
    x<-mgi[grep(paste0("^",title,"$"),mgi$mgi_symbol),"ensembl_gene_id"]
  }
  
plotCounts(cds,x,intgroup=c("salt","genotype"),returnData=T) %>% 
  ggplot(aes(x=salt, y=count)) +
  geom_point(position=position_jitter(w=0.1,h=0)) + ggtitle(paste0(x,":",title)) +
 # scale_y_log10(breaks=c(25,100,400)) + 
 # scale_y_continuous(trans="log2") + ylim(0,25000) +
  expand_limits(x=0, y = 0) +
  facet_grid(~genotype) + theme_bw()
  }

gg_plotCounts("Nr1h4")
```

## Look for genes with High Interaction Term
```{r}
resultsNames(cds)
res<-results(cds, name="genotypeMif.saltNacl")
summary(res)
resDF<-as.data.frame(res)
resDF<-resDF[!is.na(resDF$padj),]
resDF<-resDF[with(resDF,order(-log2FoldChange)),]
idx<-match(rownames(resDF),mgi$ensembl_gene_id)
resDF$mgi<-mgi[idx,"mgi_symbol"]
head(resDF)
gg_plotCounts("Slc14a1")
gg_plotCounts("Lrrn3")
tail(resDF)
gg_plotCounts("Saa3")
gg_plotCounts("Zfp750")

resDF<-resDF[with(resDF,order(padj)),]
head(resDF)
gg_plotCounts("Ppbp")

gg_plotCounts("Nr1h4")
gg_plotCounts("Pkhd1")

```

## control logfc vs hnf1b logfc
```{r}
cds2<-cds
cds2$group <- factor(paste0(cds2$genotype, cds2$salt))
design(cds2)<- ~ group
cds2 <- DESeq(cds2) 
resultsNames(cds2)

summary(res_wt<-results(cds2, contrast=c("group","ControlNacl","ControlnoSalt"),alpha=0.05))
res_wt<-as.data.frame(res_wt)
colnames(res_wt)<-paste0("WT",".",colnames(res_wt))

summary(res_hnf1b<-results(cds2, contrast=c("group","MifNacl","MifnoSalt")))
res_hnf1b<-as.data.frame(res_hnf1b)
colnames(res_hnf1b)<-paste0("hnf1b",".",colnames(res_hnf1b))

res_merged<-cbind(res_wt,res_hnf1b)
head(res_merged)
res_merged<-res_merged[res_merged$WT.padj < 0.05 & res_merged$WT.baseMean > 100,]
dim(res_merged)
head(res_merged)
res_merged<-subset(res_merged,!(is.na(WT.log2FoldChange) | is.na(hnf1b.log2FoldChange)))
idx<-match(rownames(res_merged),mgi$ensembl_gene_id)
res_merged$mgi<-mgi[idx,"mgi_symbol"]
res_merged$'Effect Size'<-abs(res_merged$hnf1b.log2FoldChange-res_merged$WT.log2FoldChange)

plot(res_merged$WT.log2FoldChange,res_merged$hnf1b.log2FoldChange,pch=16,cex=0.5,xlab="WT Salt Induced Log2FC",ylab="Hnf1b Salt Induced Log2FC",col=ifelse(res_merged$`Effect Size` > 1,"red","black"))
abline(h=0,col="red2")
#res_merged[identify(res_merged$WT.log2FoldChange,res_merged$hnf1b.log2FoldChange,labels=res_merged$mgi),]
```

## load Karam's Direct Target List
```{r}
head(dt<-read.table("hnf1b_direct_targets.txt",stringsAsFactors=F))
dt<-AnnotationDbi::select(org.Mm.eg.db,columns="ENSEMBL",keytype="ALIAS",keys=dt$V1)
sum(!is.na(dt$ENSEMBL))
res_merged$'Direct Target'<-rownames(res_merged) %in% dt$ENSEMBL
View(res_merged)

plot(res_merged$WT.log2FoldChange,res_merged$hnf1b.log2FoldChange,pch=16,cex=0.5,xlab="WT Salt Induced Log2FC",ylab="Hnf1b Salt Induced Log2FC",col=ifelse(res_merged$`Direct Target`,"red","black"))

res_merged<-res_merged[with(res_merged,order('Direct Target','Effect Size')),]
View(res_merged)
res_merged[res_merged$`Direct Target` & res_merged$`Effect Size` > 1,]

gg_plotCounts("Arhgap24")
gg_plotCounts("Col4a1")
```

```{bash MACS,eval=F}
macs2 callpeak --call-summits -c igg_rep1_SRR2124926.dedup.unique.bam  -t hnf1b_rep1_SRR2124924.dedup.unique.bam -n hnf1b_rep1 -g mm
macs2 callpeak --call-summits -c igg_rep2_SRR2124927.dedup.unique.bam  -t hnf1b_rep2_SRR2124925.dedup.unique.bam -n hnf1b_rep2 -g mm
```

##
```{r}
narrowPeakToGRanges<-function(file) {
  x <- read.table(file,stringsAsFactors=F)
  gr <-GRanges(seqnames=x$V1, ranges = IRanges(start=x$V2, end=x$V3),
               strand="*", score=x$V5, e=x$V7,summit=x$V10)
  return(gr)
}

#devtools::install_github("lawremi/HelloRanges")
hnf1b_rep1<-narrowPeakToGRanges("hnf1b_rep1_peaks.narrowPeak")
hnf1b_rep2<-narrowPeakToGRanges("hnf1b_rep2_peaks.narrowPeak")
pairs<-findOverlapPairs(hnf1b_rep1,hnf1b_rep2)
hnf1b<-pintersect(pairs)
head(hnf1b_rep2)

tss<-getAnnotation(ensembl_84,featureType="TSS")
tss<-tss[rownames(res[res$baseMean > 1,])]
hnf1b_anno <- annotatePeakInBatch(hnf1b, AnnotationData=tss, output="both", maxgap=100L)
summary(hnf1b_anno$shortestDistance)
table(hnf1b_anno$insideFeature)
p1<-hnf1b_anno[hnf1b_anno$insideFeature=="upstream" & hnf1b_anno$shortestDistance < 50000,]$feature
p2<-hnf1b_anno[hnf1b_anno$insideFeature=="inside" | hnf1b_anno$insideFeature=="overlapStart" | hnf1b_anno$insideFeature=="overlapEnd"]$feature
p<-unique(c(p1,p2))
res_merged$`Nearby Peak`<-rownames(res_merged) %in% p
sum(res_merged$`Nearby Peak`)
View(res_merged)
```

## Create Output
```{r}
colnames(res_merged)
temp<-res_merged[,c("mgi","WT.baseMean","WT.log2FoldChange","WT.padj","hnf1b.log2FoldChange","hnf1b.padj","Effect Size","Direct Target","Nearby Peak")]
colnames(temp)<-c("MGI Symbol","baseMean","WT.log2FoldChange","WT.padj","Hnf1b.log2FoldChange","Hnf1b.padj","Effect_Size","Direct_Target","Nearby Peak")
temp<-temp[with(temp,order(-Effect_Size)),]
head(temp,40)
write.csv(temp,file="hnf1b_nacl_log2FCs_wChIPdata_071016.csv",quote=F)
options(httr_oob_default=TRUE)
gs_auth(new_user = TRUE)
googlesheets::gs_upload("hnf1b_nacl_log2FCs_wChIPdata_071016.csv")
```

## GO Analysis
```{r}
expressed_genes<-rownames(res_wt[res_wt$WT.baseMean > 1 & !is.na(res_wt$WT.log2FoldChange) & !is.na(res_wt$WT.padj),])
gocat<-AnnotationDbi::select(Mus.musculus,keys=expressed_genes,keytype="ENSEMBL",columns="GOID")
gocat<-gocat[gocat$ONTOLOGY=="BP",c("ENSEMBL","GOID")]
gocat$GOID<-as.character(gocat$GOID)
gocat.list<-split(gocat$GOID,gocat$ENSEMBL)
gocat.list[["ENSMUSG00000047638"]] #fxr
gocat.list[["ENSMUSG00000030109"]] #Slc6a12
gocat.list[["ENSMUSG00000043760"]] #Pkhd1

#bias data
bd<-sum(width(reduce(ranges(genehits))))
bd["ENSMUSG00000047638"]
#bd<-bd[names(bd) %in% expressed_genes]
head(bd<-bd[expressed_genes])

#bias data
temp<-res_wt[expressed_genes,]
degs<-as.numeric(temp$WT.padj < 0.05 & abs(temp$WT.log2FoldChange ) > 1)
names(degs)<-rownames(temp)
table(degs)
summary(degs)


listGO<-function(goid) {
print(OrganismDbi::select(Mus.musculus,keys=goid,keytype="GOID",columns="TERM"))
tg<-OrganismDbi::select(Mus.musculus,keys=gocat[grep(goid,gocat$GOID),"ENSEMBL"],keytype="ENSEMBL",columns="SYMBOL")
tg$deg<-degs[tg$ENSEMBL]
tg
}

pwf<-nullp(degs,bias.data=bd)
GO.wall<-goseq(pwf,gene2cat=gocat.list)

head(GO.wall,20) %>%
  mutate(term=factor(term,levels=rev(term))) %>%
ggplot(aes(x=term,y=-log10(over_represented_pvalue))) +
  geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") +
  theme_bw() 


temp<-listGO("GO:0006814")
head(temp)
res_merged$sodium_transport<-rownames(res_merged) %in% temp$ENSEMBL
res_merged[res_merged$sodium_transport & res_merged$`Effect Size` > 1,]$mgi

temp<-listGO("GO:0006811")
head(temp)
res_merged$ion_transport<-rownames(res_merged) %in% temp$ENSEMBL
View(res_merged)
res_merged[res_merged$ion_transport & res_merged$`Effect Size` > 1,]$mgi

```


## Wt vs DN
```{r}
summary(res_genotype<-results(cds2, contrast=c("group","MifnoSalt","ControlnoSalt"),alpha=0.05))
colnames(res_genotype)<-paste0("Wt_vs_DN",".",colnames(res_genotype))
#head(res_genotype<-as.data.frame(res_genotype))
#head(res_genotype[!(is.na(res_genotype$Wt_vs_DN.padj) | is.na(res_genotype$Wt_vs_DN.log2FoldChange)),])
res_genotype<-res_genotype[!(is.na(res_genotype$Wt_vs_DN.padj) | is.na(res_genotype$Wt_vs_DN.log2FoldChange)),]
idx<-match(rownames(res_genotype),mgi$ensembl_gene_id)
res_genotype$mgi<-mgi[idx,]$mgi_symbol
res_genotype<-res_genotype[with(res_genotype,order(Wt_vs_DN.padj,Wt_vs_DN.log2FoldChange)),]
head(res_genotype)
res_genotype$`Nearby Peak`<-rownames(res_genotype) %in% p
res_genotype$'Direct Target'<-rownames(res_genotype) %in% dt$ENSEMBL

temp<-res_genotype[,c("mgi","Wt_vs_DN.baseMean","Wt_vs_DN.log2FoldChange","Wt_vs_DN.padj","Direct Target","Nearby Peak")]
head(temp)
gg_plotCounts("Pde4c")
write.csv(temp,file="Wt_vs_DN_log2FCs_wChIPdata_071416.csv",quote=F)
options(httr_oob_default=TRUE)
googlesheets::gs_auth(new_user = TRUE)
temp_gs<-googlesheets::gs_upload("Wt_vs_DN_log2FCs_wChIPdata_071416.csv")
googlesheets::gs_browse(temp_gs)

```

"Mon Aug  8 13:43:37 2016"
 
 Want to make a heatmap of the Log2FC data of genes that don't change in the Mutant but do change in the WT.  Also a Log2Fc wt vs Log2FC mutant.
  A - Schematic
  B - Heatmap with Log2FC values (2 colums, 45+27 genes, annotation of wheter or not it is a direct target or not).  Sorted by size effect, not clustered
  C - Stacked Barchart?  Number of Genes with Ups and Downs
  D - Scatterplot with LogFC WT vs LogFC MT
  
  
```{r}
head(res_merged)

nrow(temp<-res_merged[abs(res_merged$WT.log2FoldChange) > 0.5 & abs(res_merged$hnf1b.log2FoldChange) < 0.1,])
nrow(temp<-res_merged[abs(res_merged$WT.log2FoldChange) > 0.5,])
nrow(temp<-res_merged)
plot(temp$WT.log2FoldChange,temp$hnf1b.log2FoldChange,cex=0.5,pch=16,col=ifelse(temp$`Nearby Peak`,"red","black"))
identify(temp$WT.log2FoldChange,temp$hnf1b.log2FoldChange,labels=temp$mgi)
```

```{r}
ftu<-read_excel("failed_to_upregulate.xlsx")
ftu<-ftu[,1:10]
ftd<-read_excel("failed_to_downregulate.xlsx")
ftd<-ftd[,1:10]
ft<-as.data.frame(rbind(ftu,ftd))
plot(ft$WT.log2FoldChange,ft$Hnf1b.log2FoldChange,cex=0.5,pch=16,col=ifelse(ft$`Nearby Peak`,"red","black"))
identify(ft$WT.log2FoldChange,ft$Hnf1b.log2FoldChange,labels=ft$'MGI Symbol')
```

## Remade res_merged to get the signs correct
```{r}
temp<-res_merged[,c("mgi","WT.baseMean","WT.log2FoldChange","WT.padj","hnf1b.log2FoldChange","hnf1b.padj","Effect Size","Direct Target","Nearby Peak")]
colnames(temp)<-c("MGI Symbol","baseMean","WT.log2FoldChange","WT.padj","Hnf1b.log2FoldChange","Hnf1b.padj","Effect_Size","Direct_Target","Nearby Peak")
#Keep Karam's Peaks
temp<-temp[ft[,1],]
temp$`Nearby Peak`<-factor(ifelse(temp$`Nearby Peak`,"Bound by Hnf1b","Not Bound"))

dim(x<-temp[with(temp,order(-WT.log2FoldChange)),])
hist(x$WT.log2FoldChange,breaks = 100)
dim(x<-x[abs(x$WT.log2FoldChange) > 1.0,])

adf<-x[,"Nearby Peak",drop=F]
ann_colors = list( `Nearby Peak` = c("Not Bound"="white", "Bound by Hnf1b"="black") )


pheatmap(x[,c("WT.log2FoldChange","Hnf1b.log2FoldChange")],cluster_rows=F,cluster_cols=F,labels_row=x[,"MGI Symbol"],
         annotation_row = adf,annotation_colors = ann_colors,annotation_legend = FALSE)
```

## Barchart
```{r}

```

